### DISCUSSION TITLE
[QUESTION] Where to get ANGLE build that has WinUI 3 SwapChainPanel support?

### DISCUSSION THREAD

**Discussion by atomic-dreams** [mono/SkiaSharp#2941] [Q&A] [answered]
I want to perform some direct 3D rendering through Silk.NET in WinUI. I'm not using SkiaSharp directly but took some code for AngleSwapChainPanel (basically AngleSwapChainPanel, GlesContext and Egl). 

As for the ANGLE, I'm using libEGL.dll and libGLESv2.dll from the latest Chromium build, and it fails to create EGL surface.

My question is, does anybody know if that ANGLE build supports WinUI SwapChainPanel (which would mean that I've messed up some configuration), or I should build ANGLE myself for that purpose? Also, where SkiaSharp acquires that dll for its package?

**Comment by mattleibow** [comment, marked as answer]
I sent a PR into the upstream ANGLE project and now you can build your own.

Here is my build script that builds it for SkiaSharp, but you can convert that to whatever. Mostly downloading things: https://github.com/mono/SkiaSharp/blob/main/native/winui/ANGLE.cake

The build is simpler if you check out the angle code and follow the official checkout and build things, but it was annoying to do in my setup, so I just do the things I know I need manually.

I do not believe anyone has an official build of ANGLE with WinUI and I just rolled my own. You can also just download the nuget and extract the builds if you just need it: https://www.nuget.org/packages/SkiaSharp.Views.WinUI/3.0.0-preview.3.1

**Reply by atomic-dreams** [reply]
Thank you!

I was able to make this work using dlls from the package for now.

Btw, I've tried to use PropertySet methods in C#, and that causes call to ANGLE to give an access violation error, while the same things using C++ WinRT Component work fine. Why that could be happening?

**Comment by ToolmakerSteve** [comment]
@atomic-dreams 

>  I've tried to use PropertySet methods in C#

From what I read, don't use PropertySet from C#. Marshaling messes up the type. 

See [this stackoverflow Q&A](https://stackoverflow.com/q/39432377/199364).

Especially the first comment by Hans Passant, and the post linked to in that comment.

--------------------------------------------------------

Looks like OpenTK uses Angle's `libglesv2.dll` on Windows. If so, could use OpenTK as the C# wrapper. Use OpenTK methods, rather than directly using PropertySet.

Hopefully mattleibow will know better, if I am wrong about this... Or know if there is an alternative to using OpenTK.
